p2p/agent.go

```diff
@@ -349,19 +364,59 @@ func (p *agent) Start(ctx context.Context) error {
 
 	close(ready)
 
+	p.batchManager.Start()
+	ctx, cancel := context.WithCancel(context.Background())
+	go p.batchFactory(ctx)
+	p.cancelHandler = cancel
```

Comment:
having both `batchManager.Start()` and `cancelHandler` is a weird implementation.